File: 백트래킹/12100번 2048 (Easy).cpp
#include<iostream>

#define MAX 20

using namespace std;

int n, ans;
int board[MAX][MAX];
int copy_board[MAX][MAX];
int dir[5];

void input()
{
    cin >> n;

    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            cin >> board[i][j];
}

void move_board(int dir)
{
    // 위쪽으로 움직이기
    if (dir == 0) {
        for (int i = 0; i < n - 1; i++)
        {
            for (int j = 0; j < n; j++)
            {
                bool check = false;
                if (copy_board[i][j] == 0)
                {
                    int k = i + 1;
                    while (k <= n - 1)
                    {
                        if (copy_board[k][j] != 0)
                        {
                            check = true;
                            break;
                        }
                        k++;
                    }

                    if (check == true)
                    {
                        copy_board[i][j] = copy_board[k][j];
                        copy_board[k][j] = 0;
                    }
                }
            }
        }

        for (int i = 0; i < n - 1; i++)
        {
            for (int j = 0; j < n; j++)
            {
                if (copy_board[i][j] == copy_board[i + 1][j])
                {
                    copy_board[i][j] = copy_board[i][j] * 2;
                    copy_board[i + 1][j] = 0;
                }
            }
        }

        for (int i = 0; i < n - 1; i++)
        {
            for (int j = 0; j < n; j++)
            {
                bool check = false;
                if (copy_board[i][j] == 0)
                {
                    int k = i + 1;
                    while (k <= n - 1)
                    {
                        if (copy_board[k][j] != 0)
                        {
                            check = true;
                            break;
                        }
                        k++;
                    }

                    if (check == true)
                    {
                        copy_board[i][j] = copy_board[k][j];
                        copy_board[k][j] = 0;
                    }
                }
            }
        }
    }
    // 오른쪽으로 움직이기
    else if (dir == 1) {
        for (int i = 0; i < n; i++)
        {
            for (int j = n - 1; j > 0; j--)
            {
                bool check = false;
                if (copy_board[i][j] == 0)
                {
                    int k = j - 1;
                    while (k >= 0)
                    {
                        if (copy_board[i][k] != 0)
                        {
                            check = true;
                            break;
                        }
                        k--;
                    }
                    if (check == true)
                    {
                        copy_board[i][j] = copy_board[i][k];
                        copy_board[i][k] = 0;
                    }
                }
            }
        }

        for (int i = 0; i < n; i++)
        {
            for (int j = n - 1; j > 0; j--)
            {
                if (copy_board[i][j] == copy_board[i][j - 1])
                {
                    copy_board[i][j] = copy_board[i][j] * 2;
                    copy_board[i][j - 1] = 0;
                }
            }
        }

        for (int i = 0; i < n; i++)
        {
            for (int j = n - 1; j > 0; j--)
            {
                bool check = false;
                if (copy_board[i][j] == 0)
                {
                    int k = j - 1;
                    while (k >= 0)
                    {
                        if (copy_board[i][k] != 0)
                        {
                            check = true;
                            break;
                        }
                        k--;
                    }
                    if (check == true)
                    {
                        copy_board[i][j] = copy_board[i][k];
                        copy_board[i][k] = 0;
                    }
                }
            }
        }
    }
    // 아래쪽으로 움직이기
    else if (dir == 2) {
        for (int i = n - 1; i > 0; i--)
        {
            for (int j = 0; j < n; j++)
            {
                bool check = false;
                if (copy_board[i][j] == 0)
                {
                    int k = i - 1;
                    while (k >= 0)
                    {
                        if (copy_board[k][j] != 0)
                        {
                            check = true;
                            break;
                        }
                        k--;
                    }
                    if (check == true)
                    {
                        copy_board[i][j] = copy_board[k][j];
                        copy_board[k][j] = 0;
                    }
                }
            }
        }

        for (int i = n - 1; i > 0; i--)
        {
            for (int j = 0; j < n; j++)
            {
                if (copy_board[i - 1][j] == copy_board[i][j])
                {
                    copy_board[i][j] = copy_board[i][j] * 2;
                    copy_board[i - 1][j] = 0;
                }
            }
        }
        for (int i = n - 1; i > 0; i--)
        {
            for (int j = 0; j < n; j++)
            {
                bool check = false;
                if (copy_board[i][j] == 0)
                {
                    int k = i - 1;
                    while (k >= 0)
                    {
                        if (copy_board[k][j] != 0)
                        {
                            check = true;
                            break;
                        }
                        k--;
                    }
                    if (check == true)
                    {
                        copy_board[i][j] = copy_board[k][j];
                        copy_board[k][j] = 0;
                    }
                }
            }
        }
    }
    // 왼쪽으로 움직이기
    else {
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < n - 1; j++)
            {
                bool check = false;
                if (copy_board[i][j] == 0)
                {
                    int k = j + 1;
                    while (k <= n - 1)
                    {
                        if (copy_board[i][k] != 0)
                        {
                            check = true;
                            break;
                        }
                        k++;
                    }
                    if (check == true)
                    {
                        copy_board[i][j] = copy_board[i][k];
                        copy_board[i][k] = 0;
                    }
                }
            }
        }

        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < n - 1; j++)
            {
                if (copy_board[i][j] == copy_board[i][j + 1])
                {
                    copy_board[i][j] = copy_board[i][j] * 2;
                    copy_board[i][j + 1] = 0;
                }
            }
        }

        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < n - 1; j++)
            {
                bool check = false;
                if (copy_board[i][j] == 0)
                {
                    int k = j + 1;
                    while (k <= n - 1)
                    {
                        if (copy_board[i][k] != 0)
                        {
                            check = true;
                            break;
                        }
                        k++;
                    }
                    if (check == true)
                    {
                        copy_board[i][j] = copy_board[i][k];
                        copy_board[i][k] = 0;
                    }
                }
            }
        }
    }
}

void select_move_dir(int cnt)
{
    if (cnt == 5)
    {
        // 기존 보드의 형태를 유지하기 위해 복사
        for (int i = 0; i < n; ++i)
            for (int j = 0; j < n; ++j)
                copy_board[i][j] = board[i][j];

        // 백트래킹을 통해 얻은 중복순열에 따라 보드 움직이기
        for (int i = 0; i < 5; ++i)
            move_board(dir[i]);

        // 최댓값 구하기
        for (int i = 0; i < n; ++i)
            for (int j = 0; j < n; ++j)
                ans = max(ans, copy_board[i][j]);

        return;
    }

    for (int i = 0; i < 4; i++)
    {
        // 0부터 3까지(위, 오른쪽, 아래, 왼쪽)의 방향으로 이루어진 중복순열 저장
        dir[cnt] = i;
        select_move_dir(cnt + 1);
    }
}

void init()
{
    ios::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);
}

int main(void)
{
    init();
    input();
    select_move_dir(0);

    cout << ans;

    return 0;
}

File: 백트래킹/12100번 2048 (Easy).py
import sys
from copy import deepcopy

input = sys.stdin.readline

def move(board, dir):
    # 동쪽으로 이동
    if dir == 0:
        for i in range(n):  # 모든 행에 대하여 수행
            top = n - 1
            for j in range(n - 2, -1, -1):
                if board[i][j]:
                    tmp = board[i][j]
                    board[i][j] = 0
                    if board[i][top] == 0:
                        board[i][top] = tmp
                    elif board[i][top] == tmp:
                        board[i][top] *= 2
                        top -= 1
                    else:
                        top -= 1
                        board[i][top] = tmp
    # 서쪽으로 이동
    elif dir == 1:
        for i in range(n):  # 모든 행에 대하여 수행
            top = 0
            for j in range(1, n):
                if board[i][j]:
                    tmp = board[i][j]
                    board[i][j] = 0
                    if board[i][top] == 0:
                        board[i][top] = tmp
                    elif board[i][top] == tmp:
                        board[i][top] *= 2
                        top += 1
                    else:
                        top += 1
                        board[i][top] = tmp
    # 남쪽으로 이동
    elif dir == 2:
        for j in range(n):  # 모든 열에 대하여 수행
            top = n - 1
            for i in range(n - 2, -1, -1):
                if board[i][j]:
                    tmp = board[i][j]
                    board[i][j] = 0
                    if board[top][j] == 0:
                        board[top][j] = tmp
                    elif board[top][j] == tmp:
                        board[top][j] *= 2
                        top -= 1
                    else:
                        top -= 1
                        board[top][j] = tmp
    # 북쪽으로 이동
    else:
        for j in range(n):  # 모든 열에 대하여 수행
            top = 0
            for i in range(1, n):
                if board[i][j]:
                    tmp = board[i][j]
                    board[i][j] = 0
                    if board[top][j] == 0:
                        board[top][j] = tmp
                    elif board[top][j] == tmp:
                        board[top][j] *= 2
                        top += 1
                    else:
                        top += 1
                        board[top][j] = tmp
    
    return board

# 알고리즘 부분
def select_dir(board, cnt):
    global ans

    if cnt == 5:
        for i in range(n):
            for j in range(n):
                ans = max(ans, board[i][j])
        return

    for i in range(4):
        tmp_board = move(deepcopy(board), i)    # 기존 보드의 형태를 유지하기 위함
        select_dir(tmp_board, cnt + 1)

# 메인 코드 부분
n = int(input())
arr = [list(map(int, input().split())) for _ in range(n)]

ans = 0
select_dir(arr, 0)
print(ans)